### PR TITLE
[ML] Memory management for incremental training

### LINKED ISSUE #1790: [ML] Boosted tree incremental training memory estimation
The memory requirements are different to regular train and we need custom upfront memory estimation code.

### PR DESCRIPTION
**Pull request by tveasey**
We were missing memory management for the new incremental training, prediction and compute encodings tasks. This change adds it in. Note that we also need to write out the trained model size to meta data and supply this when we call tasks which load a pre-existing model to do proper accounting.

Closes #1790.

### CODE REVIEW

File: lib/maths/analytics/CBoostedTreeFactory.cc [hunk 1]
```diff
@@ -1631,14 +1642,31 @@ std::size_t CBoostedTreeFactory::estimateMemoryUsageForTrain(std::size_t numberR
...
 CBoostedTreeFactory::estimateMemoryUsageForTrainIncremental(std::size_t numberRows,
                                                             std::size_t numberColumns) const {
-    return m_TreeImpl->estimateMemoryUsageTrainIncremental(numberRows, numberColumns);
+    std::size_t maximumNumberTrees{this->mainLoopMaximumNumberTrees(
+        m_TreeImpl->m_Hyperparameters.eta().fixed()
+            ? m_TreeImpl->m_Hyperparameters.eta().value()
+            : computeEta(numberColumns))};
+    CScopeBoostedTreeParameterOverrides<std::size_t> overrides;
+    overrides.apply(m_TreeImpl->m_Hyperparameters.maximumNumberTrees(), maximumNumberTrees);
+    return m_TreeImpl->estimateMemoryUsageForTrainIncremental(numberRows, numberColumns);
```
Comment (valeriy42) [inline lib/maths/analytics/CBoostedTreeFactory.cc:1657]: Should this code be moved to m_TreeImpl?

Comment (tveasey) [reply lib/maths/analytics/CBoostedTreeFactory.cc:1657]: The tricky part is that `mainLoopMaximumNumberTrees` logic relates to `CBoostedTreeFactory`. This sort of configuration seems to naturally belong on the factory, so I'm not sure I'd want to move that. `estimateMemoryUsageForTrainIncremental` does implement the bulk of memory estimation so I feel like this is probably ok.

**Comment by tveasey** [comment]
> I wonder if we need another test for memory estimate

Yes good point. I'll add some more testing in this area.

**Comment by tveasey** [comment]
I added some more testing (and it actually found a pre-existing bug :)). Could you take another look @valeriy42?

File: lib/api/unittest/testfiles/model_meta_data/model_meta_data.schema.json [hunk 2]
```diff
@@ -112,6 +112,9 @@
                 },
                 "loss_gap" : {
                     "type" : "number"
+                },
+                "trained_model_memory_usage": {
+                    "type" : "number"
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/api/unittest/testfiles/model_meta_data/model_meta_data.schema.json:117]: Agreed. I'll add an issue so we capture this, but will make the change when we actually merge to main.

**Pull request merged**